File: day58/diameterOfBT.cpp
// 543. Diameter of Binary Tree
#include<iostream>
#include<vector>
using namespace std;


struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int dfsHeight(TreeNode* root, int& maxi){
  if(root == nullptr) return 0;
  int lh = dfsHeight(root->left, maxi);
  int rh = dfsHeight(root->right, maxi);
  
  maxi = (maxi, lh+rh);
  return 1 + max(lh, rh);
}

int maxi = 0;
int diameterOfBT(TreeNode* root){
  if(root == nullptr) return 0;
  int maxi = 0;
  dfsHeight(root, maxi);
  return maxi;
}

int main(){

   // Creating a balanced tree
  TreeNode* root = new TreeNode(1);
  root->left = new TreeNode(2);
  root->right = new TreeNode(3);
  root->left->left = new TreeNode(4);
  root->left->right = new TreeNode(5);
  root->right->left = new TreeNode(6);
  root->right->right = new TreeNode(7);

  cout << "Diameter: " << diameterOfBT(root) << endl;

  // Creating an unbalanced tree
  TreeNode* unbalancedRoot = new TreeNode(1);
  unbalancedRoot->left = new TreeNode(2);
  unbalancedRoot->left->left = new TreeNode(3);
  unbalancedRoot->left->left->left = new TreeNode(4);

  cout << "Diameter2: " << diameterOfBT(unbalancedRoot)<<endl;

  return 0;
}

File: day103/DiameterOfBT.java
// 543. Diameter of Binary Tree
public class DiameterOfBT {
  private int maxDiameter = 0;
  private int helper(TreeNode root){
    if(root==null) return 0;

    int leftHt = helper(root.left), rightHt = helper(root.right);
    maxDiameter = Math.max(maxDiameter, rightHt+leftHt);

    return Math.max(leftHt, rightHt)+1;
  }

  public int diameterOfBinaryTree(TreeNode root) {
    if(root==null) return 0;
    helper(root);
    return maxDiameter;
  }

  // Sample input and main method
  public static void main(String[] args) {
    /*
     * Sample tree:
     *      1
     *     / \
     *    2   3
     *   / \
     *  4   5
     * Diameter is 3 (path: 4-2-1-3 or 5-2-1-3)
     */
    TreeNode root = new TreeNode(1);
    root.left = new TreeNode(2);

    // TreeNode root = new TreeNode(1);
    // root.left = new TreeNode(2);
    // root.right = new TreeNode(3);
    // root.left.left = new TreeNode(4);
    // root.left.right = new TreeNode(5);

    DiameterOfBT tree = new DiameterOfBT();
    int diameter = tree.diameterOfBinaryTree(root);
    System.out.println("Diameter of Binary Tree: " + diameter);
  }
}
